Add enderpearl thrower to cause and TeleportType to context

RedNesto opened the pull request:
All of that for the MoveEntityEvent.Teleport.
Corrects an issue of #1523 
The thrower is placed at the root of the cause, I could not figure out how to put the enderpearl at the root right now.

First time I played with the CauseStackManager, so I may have made mistakes.

JBYoshi committed:
Optimize and improve experience calculations.

ImMorpheus committed:
Implement PardonUserEvent/PardonIpEvent. Close #436

vectrixdevelops committed:
Implement spawn loading methods for dimension type mixin

parlough committed:
Proposal to fix #1568

(cherry picked from commit 8ea2a3f)

ImMorpheus committed:
Implement suggested changes

ImMorpheus committed:
These shouldn't be necessary anymore

ImMorpheus committed:
Cleanup import

gabizou committed:
Bump API version for next minor API release.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Faithcaio committed:
fix Inventory#query(Class...) not accepting non Inventory classes anymore

(cherry picked from commit b781238)

Cybermaxke committed:
Improved the generic event.

This fixes generic events that have complex generic types, like DataRegistration<?,?>, Key<?>, etc. Also fixes the MyHomes plugin for testing (make it run without errors). Remove event post method on client thread, was already fixed before.

Cybermaxke committed:
Init the console permissions on the client once the server is available.

Faithcaio committed:
fixes NoSuchElementException on Advancements

(cherry picked from commit 8bc303b)

Cybermaxke committed:
Keys must be registered through the event or registry.

- Added support for non static inner classes in the typetoken helper
- The plugin id for keys will be retrieved from the current plugin container (old method is deprecated, but still working)

gabizou committed:
Attempt to improve the exception printout to better improve the stacktrace provided.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Add Move bot configuration for moving issues.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Cybermaxke committed:
Improved the TypeTokenHelper#isAssignable method.

gabizou committed:
Fix failed data not being removed after successful deserialization. Fixes #1683.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

me4502 committed:
Fixed dispenser event not firing. Fixes #1623 (#1628)

* Fixed dispenser event not firing

* Added a test for dispenser fire events. Plays an explosion sound from a dispenser when it fires.

* Added a command to enable/disable dispenser fire test.

Cybermaxke committed:
Fix that the causestackmanager was used on the client.

ImMorpheus committed:
Add test plugin

ImMorpheus committed:
No one saw me

Zidane committed:
Per cybermaxke's PR subset, fix CauseStack on client event issue

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

gabizou committed:
Add configurable purge custom data that failed to load.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Merge generic event fixes. Merges #1681.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Fix custom data config generation. Sponge now also registers a Set type serializer for Configurate.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Faithcaio committed:
implement CraftItemEvents

parlough committed:
Make the damage source test not annoying

pie-flavor committed:
Fix skipping BlockChangeFlags.NEIGHBOR_PHYSICS

pie-flavor committed:
Rename the field too

gabizou committed:
Merge PardonUserEvent implementation fix and causes fixes for ban service events.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

parlough committed:
Fix data serializers which use the month - Fixes #1317

gabizou committed:
Some minor cleanup. Attempt to attach the first block snapshot as part of the cause for some entity spawns when blocks are changed by a player. Fixes #1001.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Allow commands to be executed by plugins during event listeners.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Add an instant change block event for falling block entities to check themselves. Partially fixes #1251.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

parlough committed:
Fix copy for fall distance data

Meronat committed:
Merge pull request #1697 from Meronat/fix/month-serializers

Fix data serializers which use the month - Fixes #1317

Meronat committed:
Fix ride entity event not firing - Fixes #1345 (#1698)

Meronat committed:
Fix pre pickup event casting error - Fixes #1695 (#1696)

parlough committed:
Update API submodule for deprecations

parlough committed:
Properly fix API submodule

Zidane committed:
Have migrator "fix" save data existing in wrong folder on client.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

ImMorpheus committed:
Throw an exception when using Entity#remove on a player. Close #989 (#1704)

* Throw an exception when using Entity#remove on a player. Close #989

* Add exception message

Faithcaio committed:
fix swap item in hand cancel using the open container
closes #1705

Faithcaio committed:
attack entity events for hanging and minecarts

dualspiral committed:
Fix scoreboard not using the correct display mode in some circumstances

Fixes #1711

Commit:
Exclude testplugins from the generated pom

gabizou committed:
Merge world folder fix for clients. Fixes SpongePowered/SpongeForge#1921.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Add some fake player detection to avoid exceptions during fake player object initialization.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Add TileEntity#toString() override for better debugging in phase tracker printouts.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Faithcaio committed:
don't rely on Hotbar dictating the width of the player inventory

Faithcaio committed:
fix closing inventory not capturing items for DropItemEvent

Faithcaio committed:
fix incorrect handling of cursor transactions when all SlotTransactions are invalidated

parlough committed:
Update API submodule for various API cleanup

Faithcaio committed:
fix MainPlayerInventoryLensImpl having hotbar & grid ordered differently for InventoryPlayer and ContainerPlayer
the order is now always Grid then Hotbar

(cherry picked from commit ca28968)

Faithcaio committed:
Update SpongeCommon

Faithcaio committed:
Update SpongeAPI (really)

Faithcaio committed:
expose input and output slots of ContainerRepair

Faithcaio committed:
keep inventory for players on Death event

Faithcaio committed:
add back items on a corpse to the respawned player
then clear the corpse to prevent mods duplicating items

dualspiral committed:
Bump API for command tab completion fix

Faithcaio committed:
fix entering death phase twice

(cherry picked from commit 9fbd0c3)

bloodmc committed:
Fix hotkey item drop not being restored to slot when cancelled. Fixes #1722

(cherry picked from commit b2b1c4f)

Faithcaio committed:
do not drop experience when setting keepinventory true

(cherry picked from commit ad4fb5f)

Faithcaio committed:
fix shiftclick not fireing

Faithcaio committed:
fix CraftingGridInventory to native inventory
fixes #1716

bloodmc committed:
Fix UnsupportedOperationException when cancelling 'DropItemEvent.Pre'.

Captured items are cleared in supplier before reaching consumer.

Zidane committed:
Cleanup PhaseTracker warnings and make wrapped exceptions a lot more
clearer.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Somehow my IDE was fine with this??

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

bloodmc committed:
Fix bed respawns.

(cherry picked from commit 65d1ec2)

Faithcaio committed:
fix bedspawn getting removed in dim0

(cherry picked from commit d7871a0)

bloodmc committed:
Move overworld spawn fix to copyFrom method.

(cherry picked from commit 588fae2)

gabizou committed:
Rename MixinContainer#init to avoid illegal access errors. Fixes BuildCraft/BuildCraft#4005.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

ImMorpheus committed:
Fix xpOrbs

(cherry picked from commit 9f81636)

Faithcaio committed:
do not ignore mods setting the crafting grid
this causes the CraftItemEvent.Preview to not fire in those cases

RedNesto committed:
Add enderpearl thrower to the cause and TeleportType to the context of MoveEntityEvent.Teleport

ryantheleach commented:
Should player be root for an enderpearl? I would have expected the enderpearl.

ImMorpheus commented:
I'd move the try and the addContext to `handleDisplaceEntityTeleportEvent` (since the event is thrown there).

gabizou commented on src/main/java/org/spongepowered/common/entity/EntityUtil.java:
```diff
@@ -317,6 +317,8 @@ public static boolean isEntityDead(net.minecraft.entity.Entity entity) {
         final IPhaseState state = peek.state;
         final PhaseContext<?> context = peek.context;
 
+        Sponge.getCauseStackManager().pushCause(entityIn);
```
It doesn't seem like you're popping the cause after pushing. You need to ensure that the cause stack is stack neutral after the method is completed, so either popping the cause, or wrapping with a try with resources on a stack frame.

RedNesto replied:
Where should I pop the cause ?
I don't know if I should really pop the cause since the method is called inside a try-with-resources (in fact, I'm not event sure I should do [that](https://github.com/SpongePowered/SpongeCommon/pull/1741/files/5d9823590665615f835869725b3aa037036e7478#diff-8455601503527079e2c6f540298a089eR62))

ImMorpheus commented on src/main/java/org/spongepowered/common/mixin/core/entity/item/MixinEntityEnderPearl.java:
```diff
@@ -55,9 +59,13 @@ private boolean onEnderPearlImpact(EntityPlayerMP player) {
-        MoveEntityEvent.Teleport event = EntityUtil.handleDisplaceEntityTeleportEvent(player, this.getLocation());
-        if (event.isCancelled()) {
-            return true;
+        try(CauseStackManager.StackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
```
Spacing

RedNesto commented:
I found out that the way I am adding the player to the cause adds the player another time in the Event for the /teleport command, and I don't know how to handle this.

Faithcaio committed:
fix dimension change sometimes teleporting back to location in previous dimension

phit commented:
~wip

Faithcaio committed:
Merge branch 'fix/drop-item-pre-restore' into stable-7

ImMorpheus committed:
Fix Autosave-Message showing even if saving is disabled (#1747)

(cherry picked from commit 09040e6)

Faithcaio committed:
fix closing inventory drops not getting captured correctly when calling closeInventory in some cases

ryantheleach commented on src/main/java/org/spongepowered/common/entity/EntityUtil.java:
```diff
@@ -321,6 +321,7 @@ public static boolean isEntityDead(net.minecraft.entity.Entity entity) {
 
         MoveEntityEvent.Teleport event = SpongeEventFactory.createMoveEntityEventTeleport(Sponge.getCauseStackManager().getCurrentCause(), fromTransform, toTransform, (org.spongepowered.api.entity.Entity) entityIn);
         SpongeImpl.postEvent(event);
+        Sponge.getCauseStackManager().popCause();
```
If you can, you should use 1 cause stack frame instead of a push/pop

Basically, using a cause stack frame in a try catch is using the https://docs.oracle.com/javase/8/docs/api/java/lang/AutoCloseable.html to push and pop a Cause Stack Frame (similar to the Call Stack Frame)  https://stackoverflow.com/questions/10057443/explain-the-concept-of-a-stack-frame-in-a-nutshell

Also we much prefer the try style compared to manual push / pop's due to the possibility for error or cause stack corruption.

Commit:
Move text colour mapping registration to clinit

bloodmc committed:
Add 'DropItemEvent.Pre' support for block dispenser's. Closes #1037

* Fix ItemStackSnapshot toString.

(cherry picked from commit 9f50e8c)

ryantheleach commented on src/main/java/org/spongepowered/common/mixin/core/entity/item/MixinEntityEnderPearl.java:
```diff
@@ -55,9 +59,14 @@ private boolean onEnderPearlImpact(EntityPlayerMP player) {
+            frame.addContext(EventContextKeys.TELEPORT_TYPE, TeleportTypes.ENTITY_TELEPORT);
+            frame.addContext(EventContextKeys.THROWER, (Player) player).getCurrentCause();
+
+            MoveEntityEvent.Teleport event = EntityUtil.handleDisplaceEntityTeleportEvent(player, this.getLocation());
```
Cancelling the displace here, will also cancel the damage when it hits another entity. Not sure if this is intended. ( i know you didn't code it this way, and it was already like this though)

Faithcaio committed:
fix item drops on closing inventory

(cherry picked from commit 1e15970)

Faithcaio committed:
Use API method to close inventory so we switch the PhaseState

(cherry picked from commit f0e7dff)

Zidane committed:
Merge pull request #1743 from RedNesto/fix/teleport-types-event-contexts

Add TELEPORT_TYPE EventContext for Enderman and /teleport teleportations

RedNesto committed:
Implement LightningEvent.Pre

Commit:
Revert "Move text colour mapping registration to clinit"

This reverts commit 6cef22284a9f7235180a8ca54815a9f4c321c87b.

Commit:
Null out AdvancementList code

Faithcaio committed:
fix #1750 middle drag warnings

Faithcaio committed:
fix SpongeContextCalculator NPE on Client

Commit:
disable stupidness
fixes #1748

ImMorpheus committed:
Fix Programmatic Explosions not updating the client

Faithcaio committed:
enable CarrierContainerMixins

Zidane committed:
Revert "Null out AdvancementList code"

This reverts commit ceabaf0f91f17d701818b5795546f1975f93d018.

Faithcaio committed:
Update SpongeAPI

Zidane committed:
Fix SF client crash when an Enderman teleports.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Merge pull request #1565 from JBYoshi/xp/since-level

Optimize and improve experience calculations.

ImMorpheus committed:
Use the PotionID when creating the DataContainer. (#1712)

Uses a DataContentUpdater to update the outdated data containers
(Closes #1276)

RedNesto committed:
Implement Chunk#getHighestYAt (#1770)

ImMorpheus committed:
Initial work on translations fixes (#1728)

* Initial work on translations

* Fix bed, brewindstand, concrete, flowerpot and slabs

* Fix skull and compile error

Faithcaio committed:
fix offhand swap using wrong slot

Faithcaio committed:
try more to find a CraftingInventory.
If we still can't find it do not fire Crafting events.

Faithcaio committed:
wrong cast

Zidane committed:
Merge pull request #1756 from ImMorpheus/fix/programmatic-explosions

Fix Programmatic Explosions not updating the client

RedNesto commented:
Argh, seems I failed :/
